### PR TITLE
NodeDefDetails: cancel edits when user presses browser back button

### LINKED ISSUE #1116: Error message after deleting a Code attribute
While adding a new CODE attribute, it is not clear to a user what to do after selecting a code list (=category list) item. There is X in the upper left corner, BUT IF user clicks browser's backwards arrow, Arena jumps into Form Designer View.
Then try to DELETE this attribute: there is this error message:
image

Selecting category requires a DONE button!

### PR DESCRIPTION
**Pull request by SteRiccio**
Fixes #1116

### CODE REVIEW

File: webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js [hunk 1]
```diff
@@ -20,20 +20,33 @@ export const useNodeDefDetails = () => {
...
     // Editing a nodeDef
-    if (nodeDefUuid) {
+    if (nodeDefUuid && !State.getNodeDef(state)) {
       const nodeDefSurvey = Survey.getNodeDefByUuid(nodeDefUuid)(survey)
       const validation = Survey.getNodeDefValidation(nodeDefSurvey)(survey)
       setState(State.create({ nodeDef: nodeDefSurvey, validation }))
     }
+
+    return () => {
+      // get local state from stateRef to get it up to date in this callback
```
Comment (minotogna) [inline webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:38]: what does that mean?

File: webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js [hunk 2]
```diff
@@ -20,20 +20,33 @@ export const useNodeDefDetails = () => {
...
+      if (State.isDirty(stateRef.current)) {
+        Actions.cancelEdits({ state: stateRef.current, showConfirm: false, goBack: false })
+      }
+    }
   }, [])
 
+  useEffect(() => {
+    // update state ref on local state update
+    stateRef.current = state
+  }, [state])
```
Comment (minotogna) [inline webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:48]: why keeping a ref to the local state and updating it every time the state changes?

**Review by minotogna — CHANGES_REQUESTED** [review]

File: webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js (hunk 1 above)
Comment (SteRiccio) [reply webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:38]: the local state that you get from the "state" variable is no up to date in this callback.
The only way I've found to get an updated state there was to put it in a "ref".

File: webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js (hunk 2 above)
Comment (SteRiccio) [reply webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:48]: see comment to line 36

File: webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js (hunk 1 above)
Comment (minotogna) [reply webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:38]: of course. have you tried to pass the state to the hook input param?

Comment (SteRiccio) [reply webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:38]: ...the cancelEdits function should be called only on unmount, not every time the state changes...

Comment (minotogna) [reply webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js:38]: Do you necessarily need to pass the state to that action ?
Keeping a ref that is just a reference to the state looks a bit dirty to me

**Pull request closed without merging**